app/models/concerns/taggable.rb

```diff
@@ -16,185 +13,135 @@ module Taggable
-      errors.add(:base, "Work must have required tags.") unless self.has_required_tags?
-      throw :abort
-    end
+  private
+
+  # Take the list of tags after saving, and filter by type:
+  def tags_after_saving_of_type(klass)
+    tags_after_saving.select { |tag| tag.is_a?(klass) }
   end
 
-  # Add an error message if the user tried to add invalid tags to the work
-  def check_for_invalid_tags
-    unless self.invalid_tags.blank?
-      errors.add(:base, "The following tags are invalid: " + self.invalid_tags.collect(&:name).join(', ') + ". Please make sure that your tags are less than #{ArchiveConfig.TAG_MAX} characters long and do not contain any invalid characters.")
-      self.invalid_tags.each do |tag|
-        tag.errors.to_a.each do |error|
-          errors.add(:base, error)
-        end
-      end
-    end
+  # Split the tag string at each comma, and remove excess whitespace:
+  def split_tag_string(tag_string)
+    tag_array = if tag_string.is_a?(Array)
+                  tag_string
+                else
+                  tag_string.gsub(/\uff0c|\u3001/, ",").split(",")
+                end
+
+    tag_array.map(&:squish).reject(&:blank?)
   end
 
-  private
+  # Split the tag_string, and find/create an array of tags of the given type.
+  def parse_tags_of_type(tag_string, klass)
+    tag_names = split_tag_string(tag_string)
 
-  # Returns a string (or array) of tag names
-  def tag_category_string(category, options={})
-    return "" unless self.respond_to?(category)
-    if !self.placeholder_tags.blank?
-      tag_array = self.placeholder_tags[category] || []
-    else
-      tag_array = self.send(category)
-    end
-    tag_names = tag_array.map {|tag| tag.name}
-    if options[:return_array]
-      tag_names
-    else
-      tag_names.join(ArchiveConfig.DELIMITER_FOR_OUTPUT)
-    end
+    tag_names.map do |tag_name|
+      klass.find_or_create_by_name(tag_name)
+    end.uniq
   end
 
-  # Process a string or array of tags from any tag class
-  def parse_tags(klass, incoming_tags)
-    tags = []
-    self.invalid_tags ||= []
-    klass_symbol = klass.to_s.underscore.pluralize.to_sym
-    if incoming_tags.is_a?(String)
-      # Replace unicode full-width commas
-      tag_array = incoming_tags.gsub(/\uff0c|\u3001/, ',').split(ArchiveConfig.DELIMITER_FOR_INPUT)
-    else
-      tag_array = incoming_tags
-    end
-    tag_array.each do |string|
-      string.strip!
-      next if string.blank?
-
-      tag = if Tag::USER_DEFINED.include?(klass.to_s)
-              klass.find_or_create_by_name(string)
-            else
-              klass.find_by(name: string, canonical: true)
-            end
-      next unless tag.present? && tag.is_a?(klass)
-
-      if tag.valid?
-        tags << tag if tag.is_a?(klass)
+  # Split the tag_string, and look up the tags for each of those tag names. We
+  # ignore tag type for this one, and create UnsortedTags for any tag names
+  # that we can't find.
+  def parse_tags(tag_string)
```

Comment:
Should we keep the comment that we only use this method and create UnsortedTags for bookmarks and collections?